src/script/lua_api/l_env.cpp

```diff
@@ -880,6 +880,21 @@ int ModApiEnvMod::l_find_node_near(lua_State *L)
 	return 0;
 }
 
+static void checkArea(v3s16 &minp, v3s16 &maxp)
+{
+	auto volume = VoxelArea(minp, maxp).getVolume();
+	// Volume limit equal to 8 default mapchunks, (80 * 2) ^ 3 = 4,096,000
+	if (volume > 4096000) {
+		throw LuaError("Area volume exceeds allowed value of 4096000");
+	}
+
+	// Clamp to map range to avoid problems
+#define CLAMP(arg) core::clamp(arg, (s16)-MAX_MAP_GENERATION_LIMIT, (s16)MAX_MAP_GENERATION_LIMIT)
```

Comment:
Generation limit? Shouldn't this be the map limit?

Reply:
This is the map limit. No block are created beyond these positions regardless of any config.

Reply:
Then that variable is named misleadingly... That's for another PR to fix though.

Reply:
MAX_MAP_GENERATION_LIMIT is 31000 btw.

Reply:
I said it on IRC and I will say it again here: **Please do not re-use MAX_MAP_GENERATION_LIMIT.**

This is dangerous.

Reply:
Empirically, the actual limit of where blocks can exist without crashing the game seems to be *at least* up to x=32687.

This might cause mapgen bugs though, it is *way* too close to the actual limit.

Reply:
I have verified that minetest.find_nodes_in_area() is used in MineClone2 and Mineclonia mapgen code.

Putting the clamp limit further out (at the exact interval I specified in the bug report) avoids having bugs at the edges.

Reply:
> This is the map limit. No blocks are created beyond these positions regardless of any config.
> Then that variable is named misleadingly... That's for another PR to fix though.

Actually, @sfan5 is reusing a variable here that should not be reused for this purpose IMO.

The crash and hang I reported happen at the boundaries of s16, not at mapgen boundaries.

By making this function misbehave at mapgen boundaries, it will affect Lua mapgens.

Reply:
I have verified that the map ends at exactly -30912 and 30928 with default settings (so no mapgen limit).
There is no argument to be made here why this function must work beyond -/+31000.

Reply:
> I have verified that the map ends at exactly -30912 and 30928 with default settings (so no mapgen limit).

That is the visible part of the map. We both know that that there is more after it, presumably up until 31008.

> There is no argument to be made here why this function must work beyond -/+31000.

In vanilla client it is actually possible to have a visible world that ends at x=31007.

It still seems possible to place blocks outside of that, but it is very very crashy there.

It does not matter what was advertised, though. I do not want to make the map bigger, I just want any sketchy behavior to be far away from where the actual world is or where mapgen is doing anything.

My argument in abstract would be “it is wrong to not use the boundaries of s16, but some arbitrary other boundary that has nothing to do with the function logic or the crash encoutered *and is far too close to where actual code could legitimately invoke it and trip over itself*”.

Reply:
> It still seems possible to place blocks outside of that, but it is very very crashy there.

To clarify: It is crashy when code assumes the 31000 boundary, but (Lua?) mapgen functions try to reach beyond it.

Basically, if the mapgen stops at 31000, but a function – like minetest.find_nodes_in_area() – is called in Lua mapgens, it is not exactly safe to make it misbehave at exactly that boundary. It should misbehave far away from the boundary.

@sfan5 I think the fundamental difference is that you evaluate the function in terms of “logically, should anything be there?” and I evaluate the function in terms of “how can I avoid future bugs (e.g. crashes)”.

Of course nothing *should* be there! But a) there can be something there b) the whole bug report was made initially because some clowns found out how to trigger weird behaviour out of bounds of s16 and not out of bounds of mapgen.

So choosing 31000 seems logical, but will likely lead to further bugs in exactly the space between ±31000 and the boundaries of s16. I want to eliminate that whole category of bugs by making that space as small as possible and putting it as far as possible from legit mapgen activity.